python/ray/data/dataset.py

```diff
@@ -845,15 +846,115 @@ def groupby(self, key: "GroupKeyT") -> "GroupedDataset[T]":
+            ...         {"A": x, "B": x**2} for x in range(100)])
+            ...     .agg({
+            ...         "A": ["mean", Std(ddof=0)],
+            ...         "B": ["min", "max"],
+            ...     })
+
+        Args:
+            aggs: Aggregations to compute as a map from a data subsetter
+                (column name for Arrow dataset or callable for any dataset) to
+                the (names of the) aggregations that we wish to apply to that
+                data subset.
+
+        Returns:
+            If the input dataset is a simple dataset then the output is
+            a tuple of ``(agg1, agg2, ...)`` where each tuple element is
+            the corresponding aggregation result.
+
+            If the input dataset is an Arrow dataset then the output is
+            an ``ArrowRow`` where each column is the corresponding
+            aggregation result.
+
+            The aggregation columns are ordered in accordance with the
+            depth-first dict ordering (cols, aggs)::
+
+                {col1: [agg1, agg2], col2: [agg2, agg1]}
+
+            would yield aggregation columns::
+
+                col1_agg1, col1_agg2, col2_agg2, col2_agg1
+
+            If the dataset is empty, return ``None``.
+        """
+        aggs_ = self._build_multi_agg(aggs)
+        return self.apply_bound_aggregations(*aggs_)
+
+    aggregate = agg
+
+    def _build_multi_agg(
+            self,
+            aggs: Dict[AggregateOnT, List["Aggregation"]],
+            skip_cols: Optional[List[str]] = None) -> List[BoundAggregateFn]:
+        """Build set of aggregations for applying multiple aggregations to
+        one or more columns.
+        """
+        err_msg = ("`aggs` must be a column name or callable --> list of "
+                   f"aggregations dict; instead, we got {aggs}.")
+        if not isinstance(aggs, dict):
+            raise TypeError(err_msg)
+        if len(aggs) == 0:
+            raise ValueError("Dict of aggregations must be nonempty.")
+        if not all(
+                isinstance(agg, (str, AggregateFn)) for aggs_ in aggs.values()
+                for agg in aggs_):
+            raise TypeError(err_msg)
+
+        aggs_ = []
+        for on, on_aggs in aggs.items():
+            on = self._check_and_normalize_agg_on(on, skip_cols=skip_cols)
+            if not isinstance(on, list):
+                on = [on]
+            for agg in on_aggs:
+                if isinstance(agg, str):
+                    try:
+                        agg = AGGS_TABLE[agg]()
+                    except KeyError as e:
+                        raise ValueError(
+                            f"{e!s} is not a supported aggregation: "
+                            f"{list(AGGS_TABLE.keys())}") from None
+                for on_ in on:
+                    agg_ = BoundAggregateFn(agg, on_)
+                    aggs_.append(agg_)
+        return aggs_
+
+    def apply_bound_aggregations(self, *aggs: BoundAggregateFn) -> U:
```

Comment:
What use cases aren't covered now? Can we remove this?

Reply:
AFAIK, the only use case that isn't covered is: aggregations with custom output column names. I was planning on adding support for output column naming in `.agg()` in a future PR, since I'm less sure about that API.

The only API that I've found tenable for this is the 2-tuple `(agg, output_column_name)` aggregations API:
  ```python
  Aggregation = Union[str, AggregateFn]

  def agg(self, aggs: Dict[AggregateOnT, List[Union[Aggregation, Tuple[Aggregation, str]]]]):
      ...

  ds.agg({"A": [("mean", "A_mean"), (Std(ddof=0), "biased_stddev")]})
  ```

Reply:
I think we should support column naming here, and remove this API in this PR. It's exposing internal implementation details / terminology.